File: 2023/kotlin/src/Day15.kt
private typealias Lens = Pair<String, Int>

private fun hash(step: String) =
    step.fold(0) { acc, chr -> ((acc + chr.code) * 17) % 256 }

private fun part1(input: String) =
    input.split(",").sumOf(::hash)

private fun part2(input: String): Int {
    val boxes = mutableMapOf<Int, MutableList<Lens>>().withDefault { mutableListOf() }
    val lenses = input.split(",")

    for (lens in lenses) {
        val label = lens.takeWhile { it != '=' && it != '-' }

        val box = boxes.getValue(hash(label))
        boxes[hash(label)] = box

        if (lens.endsWith("-")) {
            box.removeAll { it.first == label }
            continue
        }

        val value = lens.takeLast(1).toInt()
        val existing = box.indexOfFirst { it.first == label }

        if (existing == -1) {
            box.add(label to value)
        } else {
            box[existing] = label to value
        }
    }

    return boxes.toList().sumOf { (boxNo, box) ->
        box.mapIndexed { slotNo , (_, value) -> (boxNo + 1) * (slotNo + 1) * value }.sum()
    }
}

fun main() {
    val input = readInput("Day15")
    part1(input).println()
    part2(input).println()
}

File: 2023/swift/AdventOfCode2023/Sources/Day15.swift
struct Day15: AdventDay {
  var input: String

  typealias Lens = (String, Int)

  func hash(_ step: String) -> Int {
    step.reduce(0) { acc, chr in
      ((acc + Int(chr.asciiValue!)) * 17) % 256
    }
  }

  func part1() -> Any {
    input.split(separator: ",").map { hash("\($0)") }.sum()
  }

  func part2() -> Any {
    var boxes = [Int: [Lens]]()
    let lenses = input.split(separator: ",")

    for lens in lenses {
      let label = "\(lens.prefix(while: { $0 != "=" && $0 != "-" }))"

      var box = boxes[hash(label), default: []]

      if lens.contains("-") {
        box.removeAll(where: { $0.0 == label })
        boxes[hash(label)] = box
        continue
      }

      let value = Int("\(lens.last!)")!

      if let existing = box.firstIndex(where: { $0.0 == label }) {
        box[existing] = (label, value)
      } else {
        box.append((label, value))
      }

      boxes[hash(label)] = box
    }

    return
      boxes
      .sum { boxNo, box in
        box
          .enumerated()
          .map { slotNo, lens in (boxNo + 1) * (slotNo + 1) * lens.1 }
          .sum()
      }
  }
}
